accommodation_service/application/accommodation_service.go

```diff
@@ -36,21 +45,114 @@ func (service *AccommodationService) Update(accommodation domain.Accommodation)
 }
 
 func (service *AccommodationService) Delete(id string) error {
-	return service.store.Delete(id)
+	objId, err := primitive.ObjectIDFromHex(id)
+	if err != nil {
+		return err
+	}
+
+	return service.store.Delete(objId)
+}
+
+func (service *AccommodationService) AddAppointment(appointment domain.CreateAppointment) error {
+	accommodation, err := service.store.GetById(appointment.Id)
+	if err != nil {
+		return err
+	}
+	err = HasOverlap(appointment, accommodation.Appointments)
+	if err != nil {
+		return err
+	}
+
+	accommodation.Appointments = append(accommodation.Appointments, domain.MakeAppointmentFromCreateAppointment(appointment))
+	return service.store.Update(*accommodation)
+}
+
+func (service *AccommodationService) UpdateAppointment(appointment domain.UpdateAppointment) error {
+	accommodation, err := service.store.GetById(appointment.OldAppointment.Id)
+	if err != nil {
+		return err
+	}
+	appointments, err := RemoveOldAppointment(appointment.OldAppointment, accommodation.Appointments)
+	if err != nil {
+		return err
+	}
+
+	err = HasOverlap(appointment.NewAppointment, accommodation.Appointments)
+	if err != nil {
+		return err
+	}
+	accommodation.Appointments = append(appointments, domain.MakeAppointmentFromCreateAppointment(appointment.NewAppointment))
+	return service.store.Update(*accommodation)
+}
+
+func RemoveOldAppointment(oldAppointment domain.CreateAppointment, appointments []domain.Appointment) ([]domain.Appointment, error) {
+	for i, entity := range appointments {
+		if isExactOverlap(oldAppointment.Interval, entity.Interval) {
+			appointments[i] = appointments[len(appointments)-1]
+			return appointments[:len(appointments)-1], nil
+		}
+	}
+	return nil, errors.New("Old Appointment is not present in this Accommodation!")
+}
+
+func isExactOverlap(interval, accInterval domain.DateInterval) bool {
+	if interval.DateFrom == accInterval.DateFrom && interval.DateTo == accInterval.DateTo {
+		return true
+	}
+	return false
+}
+
+func HasOverlap(appointment domain.CreateAppointment, accAppointments []domain.Appointment) error {
+	interval := appointment.Interval
+	for _, entity := range accAppointments {
+		err := checkOverlap(interval, entity.Interval)
+		if err != nil {
+			return err
+		}
+		err = checkOverlap(entity.Interval, interval)
```

Comment:
Why is `checkOverlap` called twice. Does it not already check every possibility?

Reply:
Yes.